examples/cifar10-cnn/cnn_batchnorm.jl

```diff
@@ -0,0 +1,123 @@
+using Knet
+include(Pkg.dir("Knet", "data", "cifar.jl"))
+
+function loaddata()
+    info("Loading CIFAR 10...")
+    xtrn, ytrn, xtst, ytst, = cifar10()
+    mn = mean(xtrn, (1,2,4))
+    xtrn = xtrn .- mn
+    xtst = xtst .- mn
+    ytrn = onehot(Int.(ytrn))
+    ytst = onehot(Int.(ytst))
+    info("Loaded CIFAR 10")
+    return (xtrn, ytrn), (xtst, ytst)
+end
+
+# TODO: make nll work and remove this encoding
+function onehot(ytrnraw, numclass=10; dtype=Float32)
+    yonehot = zeros(dtype, numclass, length(ytrnraw))
+    for (i, y) in enumerate(ytrnraw)
+        yonehot[y, i] = 1.0
+    end
+    return yonehot
+end
+
+# Model definition
+kaiming(et, h, w, i, o) = et(sqrt(2 / (w * h * o))) .* randn(et, h, w, i, o)
+
+function init_model(;et=Float32)
+    w = Any[
+        kaiming(et, 3, 3, 3, 16), bnparam(et, 16),
+        kaiming(et, 3, 3, 16, 32), bnparam(et, 32),
+        kaiming(et, 3, 3, 32, 64), bnparam(et, 64),
+        xavier(et, 100, 8 * 8 * 64), bnparam(et, 100),
+        xavier(et, 10, 100), zeros(et, 10, 1)
+    ]
+    # Initialize a BNMoments object for each batchnorm
+    m = Any[BNMoments() for i = 1:4]
+    if gpu() >= 0; w = map(KnetArray{et}, w); end
```

Comment:
How do you keep track of the alignment between w and m?

Reply:
There are 4 batchnorm layers, so there should be a `BNMoments` object for each of them. In general, it is equal to `bnparam` calls. In more complex examples, user should initialize w and m arrays. This may be inconvenient for very deep models like resnet, but I can't think any solution to this problem until we switch to the modular API.